### PR TITLE
feat: add experimental support for iterators and streams

### LINKED ISSUE #1853: update request handling in resolveHttpResponse to follow Fetch API
follow up for #1852

We should change the input to `resolveHTTPResponse` to accept Fetch API objects by default. All other handlers should then be updated to map to the Fetch API.

Right now the Fetch adapter has to adapt objects to what tRPC accepts:
https://github.com/trpc/trpc/blob/456aad652c9e98e77044012f47177a620176e009/packages/server/src/adapters/fetch/fetchRequestHandler.ts#L23-L28

<sub>[TRP-66](https://linear.app/trpc/issue/TRP-66/update-request-handling-in-resolvehttpresponse-to-follow-fetch-api)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/1853">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/1853/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### LINKED ISSUE #4477: feat: Streaming Mutations / Queries
### Describe the feature you'd like to request

Many upstream APIs (e.g. OpenAI's GPT) stream responses to the client. It would be great if TRPC supported such streaming requests.

### Describe the solution you'd like to see

I'd imagine this would look like a mutation or query where the resolver is an async generator. Example:

```typescript
publicProcedure
   .streamingMutation(async function* ({ input, ctx }) {
     for(let x=0x;<10;x++) {
       yield x;
       await sleep(100);
     }
   });
```

### Describe alternate solutions

An alternative is to set up subscriptions and map the mutation inputs to some identifier in the subscription. The problem with this approach is (a) it depends on websockets, which are difficult to support from an infrastructure perspective, (b) the race conditions between establishing a socket connection and corresponding this with mutations/queries can be miserable.

### Additional information

I'd mostly like to understand any suggestions for how to approach this and whether you'd support a PR. Creating a `streamingMutation` procedure type, for instance, is fairly invasive but may be the correct approach.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4477">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4477/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by KATT**
- Closes #4477
- Closes #1853

tl;dr is that queries and mutations can return generator functions, and you can "just" use them

https://github.com/trpc/trpc/blob/77f23db4f5b54ed97e01fd50cff98602d2b6ca27/packages/tests/server/tupleson.test.ts#L383-L399

## 🎯 Changes

Proof of concept of using tupleson to handle streaming data - tl;

- [x] Make new `resolveHTTPResponse`-function that returns a fetch `Response`
- [x] Use tupleson to serialize promises on the backend
- [x] Create a `httpTuplesonLink` that handles this type of response from the server
- [x] Make procedure as a generator function possible
- [x] Make tupleson configurable in `initTRPC` and figure out a way to reuse same configuration for both client and server (transformers will be unnecessary as we can configure them all in tson)

### CODE REVIEW

File: packages/client/src/links/httpTuplesonLink.ts
Comment (KATT) [inline packages/client/src/links/httpTuplesonLink.ts:1]: This is not done at all - just a stub

File: packages/client/src/links/httpTuplesonLink.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      getUrl,
+      getBody,
+    },
+    ac,
+  );
+  const cancel = () => ac?.abort();
+  const textDecoder = new TextDecoder();
+  const promise = responsePromise.then(async (res) => {
+    const stringIterator = mapIterable(
+      readableStreamToAsyncIterable(res.body as any),
```
Comment (KATT) [inline packages/client/src/links/httpTuplesonLink.ts:86]: Not sure what is different from our tests in tupleson, but `res.body` doesn't have `.getReader()`


@Sheraff - any idea?

File: packages/client/src/links/httpTuplesonLink.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      (v) => textDecoder.decode(v as any),
+    );
+
+    type FIXME = any;
+
+    const output = await opts.parseAsync<FIXME>(stringIterator);
```
Comment (KATT) [inline packages/client/src/links/httpTuplesonLink.ts:92]: We should probably make `parseAsync` work with readable streams 🤷

File: packages/server/src/http/resolveHTTPResponse.ts [hunk 3]
```diff
@@ -453,3 +482,174 @@ export async function resolveHTTPResponse<
...
+      procedureCalls.map((p) =>
+        p.then((result) => {
+          // transform
+          return transformTRPCResponse(router._def._config, result);
+        }),
+      ),
+    );
+
+    // TODO - maybe tupleson should return a ReadableStream??
+    const stream = asyncIterableToReadableStream(tson);
```
Comment (KATT) [inline packages/server/src/http/resolveHTTPResponse.ts:621]: We should probably make tson return a stream

File: packages/client/src/links/httpTuplesonLink.ts (hunk 1 above)
Comment (Sheraff) [reply packages/client/src/links/httpTuplesonLink.ts:86]: It depends whether it's a response from Web API or a response from Node IIRC. It's been a while so I don't really remember but i had left these comments in https://github.com/trpc/trpc/blob/main/packages/client/src/links/internals/parseJSONStream.ts

```ts
  /**
   * As given by `(await fetch(url)).body`
   */
  readableStream: NodeJS.ReadableStream | WebReadableStreamEsque;
```

```ts
  // we handle 2 different types of streams, this if where we figure out which one we have
  if ('getReader' in readableStream) {
    await readStandardChunks(readableStream, onChunk);
  } else {
    await readNodeChunks(readableStream, onChunk);
  }
```

```ts
/**
 * Handle NodeJS stream
 */
function readNodeChunks(
  stream: NodeJS.ReadableStream,
  onChunk: (chunk: Uint8Array) => void,
) {
  return new Promise<void>((resolve) => {
    stream.on('data', onChunk);
    stream.on('end', resolve);
  });
}

/**
 * Handle WebAPI stream
 */
async function readStandardChunks(
  stream: WebReadableStreamEsque,
  onChunk: (chunk: Uint8Array) => void,
) {
  const reader = stream.getReader();
  let readResult = await reader.read();
  while (!readResult.done) {
    onChunk(readResult.value);
    readResult = await reader.read();
  }
}
```

Comment (KATT) [reply packages/client/src/links/httpTuplesonLink.ts]: Nice, the nodejs one is an async iterable so i could just do `for await (....)`

File: packages/server/src/adapters/node-http/nodeHTTPRequestHandler.ts [hunk 4]
```diff
@@ -123,14 +79,24 @@ export async function nodeHTTPRequestHandler<
...
+
+    if (!opts.res.statusCode || opts.res.statusCode === 200) {
+      opts.res.statusCode = res.status;
+    }
+    opts.res.statusMessage = res.statusText;
+    for (const [key, value] of res.headers.entries()) {
+      opts.res.setHeader(key, value);
+    }
+
+    if (res.body) {
```
Comment (Akxe) [inline packages/server/src/adapters/node-http/nodeHTTPRequestHandler.ts:94]: A check of `Symbol.asyncIterator in res.body`, would at least assert that the body is an `AsyncIterable`.

File: packages/server/src/http/internals/types.ts [hunk 5]
```diff
@@ -9,8 +9,31 @@ import { TRPCResponse } from '../../rpc';
...
 
+/**
+ * @deprecated use `Headers`
+ */
 export type HTTPHeaders = Dict<string[] | string>;
 
+export function httpHeadersToFetchHeaders(headers: HTTPHeaders): Headers {
+  const newHeaders = new Headers();
+  for (const [key, value] of Object.entries(headers)) {
+    if (!value) {
```
Comment (Akxe) [inline packages/server/src/http/internals/types.ts:20]: I am not sure about this check... what if someone uses the presence of a header rather than its value... people are wierd

File: packages/tests/server/tupleson.test.ts [hunk 6]
```diff
@@ -402,6 +403,65 @@ describe('with transformer', () => {
...
+describe.todo('with tupleson transformer', () => {
+  const orderedResults: number[] = [];
+  const tuplesonOptions: TsonOptions = {
+    nonce: () => '__tson',
+    types: [tsonBigint],
+  };
+  const ctx = konn()
+    .beforeEach(() => {
+      const t = initTRPC.create({
+        transformer: 'tupleson', // <------- help, idk if this is the right thing
```
Comment (Akxe) [inline packages/tests/server/tupleson.test.ts:415]: To me, it depends, if `tupleson` is the default anyway, the string is fine, if not I would opt for a reference from the `tupleson` package in order to make it tree shakable

The tree shake will, however, only fork if the default is different from `tupleson`, or if this is a required property and thus none is the default...

Comment (helmturner) [reply packages/tests/server/tupleson.test.ts:415]: > To me, it depends, if `tupleson` is the default anyway, the string is fine, if not I would opt for a reference from the `tupleson` package in order to make it tree shakable
> 
> The tree shake will, however, only fork if the default is different from `tupleson`, or if this is a required property and thus none is the default...

@Akxe
We've been leaning towards making Tupleson *the* transformer moving forward, deprecating the need for this field. It would make TSON a hard requirement for future implementations (but we maintain backward compat for a bit).

If users want to customize the transform logic, they do so by configuring TSON (which allows an alternative to the default `JSON` from stringifying/parsing).

Essentially, we would be moving the 'transformer' concept to be an implementation detail of Tupleson, instead of an implementation detail of tRPC.

That was an idea, at least. Thoughts?

Comment (Akxe) [reply packages/tests/server/tupleson.test.ts:415]: I do like the idea of using TSON. My point was merely informative :)

Comment (Akxe) [reply packages/tests/server/tupleson.test.ts:415]: I do not want a string to be there. It means that even if you were to replace it, the TSON dependency is forced upon the consumer. 

From my point of view; no choice or choice in the form of a `function` or a `Class`.

Why string? What would you put there otherwise?

**Comment by XantreGodlike** [comment]
@KATT What are the progress on it? I think streaming is really useful

**Comment by KATT** [comment]
> @KATT What is a progress on it? I think streaming is really useful

@XantreGodlike it's not high on the priority list for us right now as we don't _need it_ ourselves. Our main priority at the moment is finishing the breaking changes we have for v11.

If you work for a company that would like to fund the development of this, we may reconsider, feel free to contact us if so.

**Comment by XantreGodlike** [comment]
> > @KATT What is a progress on it? I think streaming is really useful
> 
> @XantreGodlike it's not high on the priority list for us right now as we don't _need it_ ourselves. Our main priority at the moment is finishing the breaking changes we have for v11.
> 
> If you work for a company that would like to fund the development of this, we may reconsider, feel free to contact us if so.

Nope, it's for hobby purposes

**Comment by JonathonRP** [comment]
@KATT any progress on this vs unstable_streamingLink?

**Comment by rdewolff** [comment]
Looking at the stream support 👀

**Comment by lacherogwu** [comment]
Looking forward for this as its very useful for download big files

**Comment by hansoksendahl** [comment]
Looking forward to this for LLM applications so I can stream the LLM output without having to write a separate end-point through the framework I am using.

**Comment by lrwlf** [comment]
Is there anything we can help? Looking forward to be merged soon
